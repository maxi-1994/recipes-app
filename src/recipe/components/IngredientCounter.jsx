

export const IngredientCounter = () => {

    return (
        <h1>IngredientCounter</h1>
    )
}