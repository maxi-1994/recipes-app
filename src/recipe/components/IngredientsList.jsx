import PropTypes from 'prop-types';


export const IngredientsList = ({ ingredientsList }) => {

    return (
        <>
            {
                ingredientsList.length > 0 ?
                (
                    <div className="ingredients-wrapper">
                        <h5>Ingredientes:</h5>

                        <ul className="ingredients-list">
                        {
                            ingredientsList.map( ingredient => (
                                <li key={ ingredient.name }>{ ingredient.name }</li>
                            ))
                        }
                        </ul>
                    </div>
                ) :
                (
                    <div className="alert alert-warning mt-3" role="alert">
                        Esta receta no tiene ingredientes cargados
                    </div>
                )
            }
        </>
    )
}

IngredientsList.propTypes = {
    ingredientsList: PropTypes.array.isRequired, 
};