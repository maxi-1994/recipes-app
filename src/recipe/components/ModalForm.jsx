import PropTypes from 'prop-types';
import Button from 'react-bootstrap/Button';
import Modal from 'react-bootstrap/Modal';

import { useForm } from '../../hooks/useForm';


export const ModalForm = ({ modalState, onCloseModal, onFormState }) => {

    const { formState, onInputValueChange } = useForm({
        name: '',
        description: '',
        ingredients: [],
        imagePath: '',
    });

    const onSubmitModalForm = (e) => {
        e.preventDefault();
        onFormState(formState);
        onCloseModal(false);
    }


    return (
        <Modal show={ modalState } onHide={ () => onCloseModal(false) }>
            <Modal.Header closeButton>
                <Modal.Title>Agregar nueva receta</Modal.Title>
            </Modal.Header>
            <Modal.Body>
                <form className="add-recipe-form" onSubmit={ onSubmitModalForm }>
                    <div className="form-group">
                        <label htmlFor="name">Nombre</label>
                        <input 
                            type="text" 
                            className="form-control" 
                            name="name" 
                            placeholder="Nombre de la receta"
                            value={ formState.name }
                            onChange={ onInputValueChange }
                        />
                    </div>
                    <div className="form-group">
                        <label htmlFor="description">Descripción</label>
                        <input 
                            type="text" 
                            className="form-control" 
                            name="description" 
                            placeholder="Descripción de la receta"
                            value={ formState.description }
                            onChange={ onInputValueChange }
                        />
                    </div>
                    {/* <div className="form-group">
                        <label htmlFor="ingredients">Ingredientes</label>
                        <input 
                            type="text" 
                            className="form-control" 
                            name="ingredients" 
                            placeholder="Ingredientes"
                        />
                    </div> */}
                    <div className="form-group">
                        <label htmlFor="imagePath">Ruta de la imagen</label>
                        <input 
                            type="text" 
                            className="form-control" 
                            name="imagePath" 
                            placeholder="Ruta de la imagen"
                            value={ formState.imagePath }
                            onChange={ onInputValueChange }
                        />
                    </div>
                    <Button type="submit" variant="primary">
                        Agregar
                    </Button>
                </form>
            </Modal.Body>
        </Modal>
    )
}

ModalForm.propTypes = {
    modalState: PropTypes.bool.isRequired, 
    onCloseModal: PropTypes.func.isRequired, 
    onFormState: PropTypes.func.isRequired,
};