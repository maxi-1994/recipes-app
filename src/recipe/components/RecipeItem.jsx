import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

export const RecipeItem = ({ _id, imagePath, name, description, ingredients }) => {

    const onDeleteRecipe = () => {
        console.log('onDeleteRecipe');
    }

    return (
        <div className="card">
            <img src={ imagePath } className="card-img-top" alt={ name } />
            <div className="card-body">
                <h5 className="card-title">{ name }</h5>
                <p className="card-text">{ description }</p>

                <h6>Ingredientes</h6>
                {
                    ingredients && ingredients.lenght > 0 ?
                    (
                        <ul>
                            <li>{ ingredients }</li>
                        </ul>
                    ) :
                    (
                        <p className="card-text">Esta receta no tiene ingredientes cargados</p>
                    )
                }
            </div>
            <div className="card-footer d-flex justify-content-between">
                <Link className="btn btn-primary" to={`/receta/${_id}`}><i className="bi bi-card-text"></i> Detalles</Link>
                <button className="btn btn-danger" onClick={ onDeleteRecipe }><i className="bi bi-trash"></i> Eliminar</button>
            </div>
        </div>
    )
}

RecipeItem.propTypes = {
    _id: PropTypes.string.isRequired,
    imagePath: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    ingredients: PropTypes.array.isRequired,
}