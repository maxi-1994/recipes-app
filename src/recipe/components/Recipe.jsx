import { useParams, useNavigate } from 'react-router-dom';
import { useContext, useState } from 'react';

import { AuthContext } from '../../auth/context/AuthContext';
import { RecipeContext } from '../context/RecipeContext';
import { useForm } from '../../hooks/useForm';
import { IngredientsList } from './IngredientsList';


export const Recipe = () => {

    {/*
        TODO:
        - Generar funcionalidad para mostrar u ocultar formulario de edit
        - agregar card a la lista de items con boton borrar
        - no mostrar el componentn "ingredientList" cuando se edita
        - generar el submit + generar la llamada por fetch en useFetchRecipes
        - Ver de generar un custom hook para la funcionalidad de ingredientes o para todo el edit
    */}

    const { editRecipe } = useContext(RecipeContext);

    const { formState, onInputValueChange, onResetForm } = useForm({
        name: '',
        description: '',
        ingredient: '',
    });
    const [ ingredients, setIngredients ] = useState([]);

    const navigate = useNavigate();

    const { authState } = useContext(AuthContext);
    const { user } = authState;

    const recipeListStorage = JSON.parse(localStorage.getItem('recipeList'));

    const { recipeId } = useParams();
    const recipeSelected = recipeListStorage.find(item => item._id === recipeId);

    const onBack = () => {
        navigate('/', {
            replace: false
        })
    };

    const onShowEditForm = () => {
        console.log('onShowEditForm');
        // Mostrar form para editar
    }

    const onAddIngredient = () => {
        const ingredientItem = { name: formState.ingredient.trim() };
        const newIngredient = [ ...ingredients, ingredientItem ];
        setIngredients(newIngredient);
        formState.ingredient = '';
    }

    // const onDeleteIngredient = () => {
    //     console.log('onDeleteIngredient');
    // }

    const onSubmitEditForm = (e) => {
        e.preventDefault();

        const formBody = {
            _id: recipeId,
            description: formState.description,
            imagePath: 'https://www.rionegro.com.ar/wp-content/uploads/2019/05/locro-criollo-abundante-D_NQ_NP_982225-MLA27393397210_052018-F.jpg',
            ingredients: ingredients,
            name: formState.name,
            userEmail: user.email,
        };
        
        editRecipe(formBody);
        onResetForm();
    }
    
    return (
        <>
            <div id="recipe">
                <div className="buttons-wrapper">
                    <button type="button" className="btn btn-secondary" onClick={ onBack }>
                        <i className="bi bi-arrow-left-square"></i> Volver
                    </button>

                    <button type="button" className="btn btn-primary" onClick={ onShowEditForm }>
                        <i className="bi bi-pencil-square"></i> Editar
                    </button>
                </div>
                <form onSubmit={ onSubmitEditForm }>
                    <div className="recipe-wrapper">
                        <div className="recipe-img-wrapper text-center">
                            <img className="img-thumbnail" src={ recipeSelected.imagePath } alt={ recipeSelected.name } />
                        </div>
                        <div className="recipe-details">
                            <h3>{ recipeSelected.name }</h3>
                                <input 
                                type="text" 
                                id="name"
                                name="name"
                                className="form-control" 
                                placeholder="Nombre de la receta"
                                value={ formState.name }
                                onChange={ onInputValueChange }
                            />

                            <p>{ recipeSelected.description }</p>
                            <textarea 
                                id="description"
                                name="description"
                                className="form-control" 
                                rows="3" 
                                placeholder="DescripciÃ³n"
                                value={ formState.description }
                                onChange={ onInputValueChange }
                            />
                        </div>
                    </div>

                    <div className='d-flex justify-content-between'>
                        <input 
                            type="text"
                            id="ingredient"
                            name="ingredient"
                            className="form-control"
                            placeholder="Ingrediente"
                            value={ formState.ingredient }
                            onChange={ onInputValueChange }
                        />
                        <button 
                            type="button" 
                            className="btn btn-primary" 
                            onClick={ onAddIngredient }
                            disabled={ formState.ingredient === '' }
                        >
                            Agregar ingrediente
                        </button>
                    </div>
                    
                    {/* Generar cards para cada item con boton para eliminar -> onDeleteIngredient */}
                    <ul>
                        {
                            ingredients.map(item => (
                                <li key={item.name}>{item.name}</li>
                            ))
                        }
                    </ul>

                    {/* No mostrar en el editar, mostrar solo la lista que voy llenando, cuando se submitea le pega al servicio y actualiza el estado */}
                    <IngredientsList ingredientsList={ recipeSelected.ingredients } />

                    <div>
                        <button type="submit" className="btn btn-primary">Guardar cambios</button>
                    </div>
                </form>
            </div>
        </>
    )
}