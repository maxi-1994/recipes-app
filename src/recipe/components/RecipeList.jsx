import { useContext } from 'react';
import { ToastContainer } from 'react-toastify';
import Button from 'react-bootstrap/Button';
import 'react-toastify/dist/ReactToastify.css';

import { RecipeContext } from '../context/RecipeContext';
import { useModal } from '../../hooks/useModal';
import { ModalForm } from './ModalForm';
import { RecipeItem } from './RecipeItem';

export const RecipeList = () => {
    
    // TODO -> Fixear spinner loading

    const { show, handleClose, handleShow } = useModal();

    const { recipeList, loading, addNewRecipe, deleteRecipe } = useContext(RecipeContext);

    return (
        <>
            <ToastContainer />

            <h2>Listado de recetas</h2>
            <Button variant="primary" onClick={ handleShow }>Agregar Receta</Button>
            <hr />

            {
                loading && (
                    <div>Cargando. . .</div>
                    // <lottie-player src="https://assets8.lottiefiles.com/packages/lf20_d2yblndy.json" background="transparent"  speed="1"  style="width: 300px; height: 300px;" loop controls autoplay></lottie-player>
                )
            }

            <div id="cards-wrapper" className="row row-cols-1 row-cols-md-3 g-4">
                {
                    recipeList.map((recipe) => (
                        <RecipeItem 
                            key={ recipe._id } 
                            recipeItems={ recipe } 
                            onDeleteRecipe={ (e) => deleteRecipe(e) } 
                        />
                    ))
                }
            </div>

            <ModalForm 
                modalState={ show } 
                onCloseModal={ (e) => handleClose(e) } 
                onFormState={ (e) => addNewRecipe(e) } 
            />
        </>
    )
}
