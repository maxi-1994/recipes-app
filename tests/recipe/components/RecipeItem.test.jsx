/* eslint-disable no-undef */
import { render } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';
import { RecipeItem } from '../../../src/recipe/components/RecipeItem';


const mockedUseContext = jest.fn();

jest.mock('react', () => ({
    ...jest.requireActual('react'),
    useContext: () => mockedUseContext
}));

describe('Pruebas en <RecipeItem />', () => {

    const recipe = {
        _id: '1',
        imagePath: 'https://www.gourmet.cl/wp-content/uploads/2022/05/Pollo-al-Horno-Web-570x458.jpg',
        name: 'pollo',
        description: 'receta pollo',
    };

    test('debe hacer match con el snapshot', () => {
        const { container } = render(<RecipeItem recipeItems={ recipe } onDeleteRecipe={() => {}} />);
        expect( container ).toMatchSnapshot();
    });
    
});