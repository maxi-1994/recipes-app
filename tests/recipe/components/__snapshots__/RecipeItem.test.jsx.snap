// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <RecipeItem /> debe hacer match con el snapshot 1`] = `
<div>
  <div
    class="col"
  >
    <div
      class="card"
    >
      <img
        alt="pollo"
        class="card-img-top"
        src="https://www.gourmet.cl/wp-content/uploads/2022/05/Pollo-al-Horno-Web-570x458.jpg"
      />
      <div
        class="card-body"
      >
        <h5
          class="card-title"
        >
          pollo
        </h5>
        <p
          class="card-text"
        >
          receta pollo
        </p>
      </div>
      <div
        class="card-footer d-flex justify-content-between"
      >
        <button
          class="btn btn-danger"
          type="button"
        >
          <i
            class="bi bi-trash"
          />
           Eliminar
        </button>
      </div>
    </div>
  </div>
</div>
`;
